northstar/src/runtime/process/init.rs

```diff
@@ -30,11 +30,130 @@ pub(super) struct Init {
+    let exit_status_tx = tx;
+    let reaper_thread = std::thread::spawn(move || {
+        loop {
+            match waitpid(None, None) {
+                Ok(WaitStatus::Exited(pid, status)) => {
+                    exit_status_tx
+                        .send(Payload::ChildExit {
+                            pid: pid.as_raw(),
+                            exit_status: ExitStatus::Exit(status),
+                        })
+                        .unwrap();
+                }
+                Ok(WaitStatus::Signaled(pid, status, _)) => {
+                    exit_status_tx
+                        .send(Payload::ChildExit {
+                            pid: pid.as_raw() as i32,
+                            exit_status: ExitStatus::Signalled(status as u8),
+                        })
+                        .unwrap();
+                }
+                Ok(WaitStatus::Continued(_)) | Ok(WaitStatus::Stopped(_, _)) => continue,
+                Err(Errno::ECHILD) => {
+                    // wait till a child process is spawned
+                    // unparked by the consumer thread after a Exec is received
+                    std::thread::park();
+                }
+                e => panic!("Failed to waitpid: {:?}", e),
+            }
+        }
+    });
+
+    // Consume messages from threads
+    loop {
+        match rx.recv() {
+            Ok(Payload::ChildExit { pid, exit_status }) => {
+                rt_tx
+                    .send(InitMessage::DeadChild { pid, exit_status })
+                    .expect("Failed to send exit status");
+
+                // Note:
+                // Currently the init process exists after the first child
+                // process exits.
+                exit(0);
+            }
+            Ok(Payload::Command(InitApi::Exec { path, args, env })) => {
+                // Start new process inside the container
+                let child_pid = fork_process(|| {
+                    // Set seccomp filter
+                    if let Some(ref filter) = seccomp {
+                        filter.apply().expect("Failed to apply seccomp filter.");
+                    }
+
+                    // Checkpoint fds are FD_CLOEXEC and act as a signal for the launcher that this child
+                    // is started. Therefore no explicit drop (close) of _checkpoint_notify is needed here.
+                    panic!(
+                        "Execve: {:?} {:?}: {:?}",
+                        &path,
+                        &args,
+                        unistd::execve(&path, &args, &env)
+                    )
+                })
+                .expect("Failed to spawn child process");
+
+                // wake up the reaper thread
+                reaper_thread.thread().unpark();
+
+                rt_tx
+                    .send(InitMessage::NewChild {
+                        pid: child_pid.as_raw(),
+                    })
+                    .expect("Failed to send exit status");
```

Comment:
Wakeup the reaper *after* sending the result to the child. If the child immediatel exits there's a chance that the runtime gets a `DeadChild` exit before a `NewChild`

Reply:
You are right! good catch